File: bluefruitconnect/Utils/Updater/DataDownloader.m
//
//  DataDownloader.m
//  Bluefruit Connect
//
//  Created by Antonio García on 31/05/16.
//  Copyright © 2016 Adafruit. All rights reserved.
//

#import "DataDownloader.h"
#import "LogHelper.h"

@implementation DataDownloader

+ (void)downloadDataFromURL:(NSURL *)url withCompletionHandler:(void (^)(NSData *))completionHandler
{
    if ([url.scheme isEqualToString:@"file"])        // Check if url is local and just open the file
    {
        NSData *data = [NSData dataWithContentsOfURL:url];
        completionHandler(data);
    }
    else
    {
        // If the url is not local, download the file
        NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration ephemeralSessionConfiguration];
        NSURLSession *session = [NSURLSession sessionWithConfiguration:configuration];
        
        NSURLSessionDataTask *task = [session dataTaskWithURL:url completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
            if (error != nil) {
                // If any error occurs then just display its description on the console.
                DLog(@"%@", [error description]);
                data = nil;
                
                completionHandler(data);
            }
            else{
                NSInteger statusCode = [(NSHTTPURLResponse *)response statusCode];
                if (statusCode != 200) {
                    DLog(@"Download file HTTP status code = %ld", (long)statusCode);
                    data = nil;
                }
                
                // Call the completion handler with the returned data on the main thread.
                if (completionHandler) {
                    [[NSOperationQueue mainQueue] addOperationWithBlock:^{
                        completionHandler(data);
                    }];
                }
            }
        }];
        
        [task resume];
    }
}

@end

File: Bluefruit/Common/AdafruitKit/Utils/DataDownloader.swift
//
//  DataDownloader.swift
//  Bluefruit
//
//  Created by Antonio on 27/01/2017.
//  Copyright © 2017 Adafruit. All rights reserved.
//

import Foundation

func downloadData(from url: URL, completion:@escaping ((Data?) -> Void)) {
    if url.scheme == "file" {       // Check if url is local and just open the file
        let data = try? Data(contentsOf: url)
        completion(data)
    } else {                          // If the url is not local, download the file
        let session = URLSession(configuration: URLSessionConfiguration.ephemeral)
        let task = session.dataTask(with: url, completionHandler: { (data, response, error) in
            guard error == nil else {
                // If any error occurs then just display its description on the console.
                DLog("Error: \(error.debugDescription)")
                completion(nil)
                return
            }

            guard let httpResponse = response as? HTTPURLResponse else {
                DLog("Download file HTTP error")
                completion(nil)
                return

            }

            let statusCode = httpResponse.statusCode
            guard statusCode == 200 else {
                DLog("Download file HTTP status code = \(statusCode)")
                completion(nil)
                return
            }

            // Call the completion handler with the returned data on the main thread
            DispatchQueue.main.async {
                completion(data)
            }
        })

        task.resume()
    }

}
